### PR TITLE
feat(server, web): smart merge

### LINKED ISSUE #5848: [BUG] Number of assets for person incorrect on merge
### The bug

When merging people the asset count of the original unnamed person is displayed rather than the total asset count for the merged person.

### The OS that Immich Server is running on

Linux Mint

### Version of Immich Server

v1.91.3

### Version of Immich Mobile App

v1.91.0

### Platform with the issue

- [ ] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
# swarm deployment extension
x-deploy-swarm: &deploy_swarm
  mode: replicated
  replicas: 1
  update_config:
    parallelism: 1
    delay: 10s
    order: stop-first

# immich environment
x-immich-env: &immich_environment
  NODE_ENV: production
  DB_HOSTNAME: db
  DB_DATABASE_NAME: ${IMMICH_DB:-immich}
  DB_USERNAME: ${IMMICH_DB_USERNAME:-immich}
  DB_PASSWORD: ${IMMICH_DB_PASSWORD:-immich}

services:
  app:
    image: ghcr.io/immich-app/immich-server:v1.91.3
    command: start-server.sh
    environment:
      <<: *immich_environment
    volumes:
      - /media/pictures/immich:/usr/src/app/upload/library
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/upload:/usr/src/app/upload/upload
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/profile:/usr/src/app/upload/profile
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/thumbs:/usr/src/app/upload/thumbs
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/encoded-video:/usr/src/app/upload/encoded-video
      - /media/pictures:/media/pictures:ro
    networks:
      - frontend
      - database
      - machine-learning
      - redis
    deploy:
      <<: *deploy_swarm
      resources:
        limits:
          cpus: ${IMMICH_APP_CPU_LIMIT}
          memory: ${IMMICH_APP_MEM_LIMIT}
      restart_policy:
        delay: 120s
        max_attempts: 3
      placement:
        constraints:
          - node.hostname == McMint

  microservices:
    image: ghcr.io/immich-app/immich-server:v1.91.3
    command: start-microservices.sh
    environment:
      <<: *immich_environment
    volumes:
      - /media/pictures/immich:/usr/src/app/upload/library
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/upload:/usr/src/app/upload/upload
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/profile:/usr/src/app/upload/profile
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/thumbs:/usr/src/app/upload/thumbs
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/encoded-video:/usr/src/app/upload/encoded-video
      - /media/pictures:/media/pictures:ro
    networks:
      - database
      - machine-learning
      - redis
    deploy:
      <<: *deploy_swarm
      resources:
        limits:
          cpus: ${IMMICH_MICROSERVICES_CPU_LIMIT}
          memory: ${IMMICH_MICROSERVICES_MEM_LIMIT}
      restart_policy:
        delay: 120s
        max_attempts: 3
      placement:
        constraints:
          - node.hostname == McMint

  machine-learning:
    image: ghcr.io/immich-app/immich-machine-learning:v1.91.3
    volumes:
      - ${STORAGE_PATH}/immich/ml-cache:/cache
    networks:
      - machine-learning
    deploy:
      <<: *deploy_swarm
      placement:
        constraints:
          - node.platform.arch == x86_64
      resources:
        limits:
          cpus: ${IMMICH_ML_CPU_LIMIT}
          memory: ${IMMICH_ML_MEM_LIMIT}

  db:
    image: tensorchord/pgvecto-rs:pg14-v0.1.11
    environment:
      POSTGRES_DB: ${IMMICH_DB:-immich}
      POSTGRES_USER: ${IMMICH_DB_USERNAME:-immich}
      POSTGRES_PASSWORD: ${IMMICH_DB_PASSWORD:-immich}
    volumes:
      - ${STORAGE_PATH}/{{index .Service.Labels "com.docker.stack.namespace"}}/db:/var/lib/postgresql/data
    networks:
      - database
    deploy:
      <<: *deploy_swarm
      resources:
        limits:
          cpus: ${IMMICH_DB_CPU_LIMIT}
          memory: ${IMMICH_DB_MEM_LIMIT}

  redis:
    image: redis:6.2.14
    networks:
      - redis
    deploy:
      <<: *deploy_swarm
      resources:
        limits:
          cpus: '0.1'
          memory: 32M

networks:
  frontend:
    external: true
    name: core_frontend
  database:
    driver: overlay
  machine-learning:
    driver: overlay
  redis:
    driver: overlay
```

### Your .env content

```Shell
N/A
```

### Reproduction steps

```bash
1. Navigate to People
2. Select any person without a name (note number of assets)
3. Type in an existing name and press done
4. In the merge dialog press yes to confirm
5. Page for person that assets were merged into loads with an asset count below name, this count is only from the previous person and not the total count
6. Refresh page to see correct count
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by martabal**
## Changes made in this PR

Currently, when you merge person A with person B, you lose the information you set about person A. With this PR, when you merge person A with person B, the server updates the person's information. person B with person A's information if person B's information is undefined. This only works when you only merge one person with another
person.

fixes #5848

### CODE REVIEW

**Comment by alextran1502** [comment]
From a user perspective, that smart merge toggle isn't clear of what it would do. To improve it, we could make this into a default mechanism.

In the case users are confused about what information about the merger is kept, we can create an info card to show the end result i.e name, birthday, ...etc

File: server/src/domain/person/person.service.ts [hunk 1]
```diff
@@ -442,46 +443,59 @@ export class PersonService {
...
+    let primaryPerson = await this.findOrFail(id);
     const primaryName = primaryPerson.name || primaryPerson.id;
 
     const results: BulkIdResponseDto[] = [];
 
     const allowedIds = await this.access.checkAccess(auth, Permission.PERSON_MERGE, mergeIds);
 
-    for (const mergeId of mergeIds) {
-      const hasAccess = allowedIds.has(mergeId);
+    for (let i = 0; i < mergeIds.length; i++) {
```
Comment (danieldietzler) [inline server/src/domain/person/person.service.ts:459]: Why did you change this from being a for each? I don't see any use of an index

File: web/src/routes/(user)/people/+page.svelte [hunk 2]
```diff
@@ -159,10 +159,19 @@
...
         id: personMerge2.id,
         mergePersonDto: { ids: [personToMerge.id] },
       });
+      if (data.person.birthDate != personMerge2.birthDate) {
+        for (const person of people) {
+          if (person.id === data.person.id) {
+            person.birthDate = data.person.birthDate;
+            break;
+          }
+        }
```
Comment (danieldietzler) [inline web/src/routes/(user)/people/+page.svelte:172]: Why is the frontend updating person information (birth date)?

File: server/src/domain/person/person.service.ts (hunk 1 above)
Comment (martabal) [reply server/src/domain/person/person.service.ts:459]: You want this smart merge to apply only for the first person and if you only merge one person

File: web/src/routes/(user)/people/+page.svelte (hunk 2 above)
Comment (martabal) [reply web/src/routes/(user)/people/+page.svelte]: Because when you merge one person with his birthDate set to a person without the birthDate set, you want to keep the birthDate

File: server/src/domain/person/person.service.ts (hunk 1 above)
Comment (danieldietzler) [reply server/src/domain/person/person.service.ts:459]: Oh I see, I've overlooked 472. My bad!

Comment (danieldietzler) [reply server/src/domain/person/person.service.ts:459]: I think you could use `for (let i in mergeIds` though

File: web/src/routes/(user)/people/+page.svelte (hunk 2 above)
Comment (danieldietzler) [reply web/src/routes/(user)/people/+page.svelte]: Yeah I get the usage but shouldn't that be taken care of by the merge on the server side?

Comment (martabal) [reply web/src/routes/(user)/people/+page.svelte]: Yep fixed

File: server/src/domain/person/person.service.ts (hunk 1 above)
Comment (martabal) [reply server/src/domain/person/person.service.ts:459]: How to get the index then ? It feels easier to use `for (let i = 0; i < mergeIds.length; i++)` than `for (let i in mergeIds)` when you need the index

Comment (danieldietzler) [reply server/src/domain/person/person.service.ts:459]: `for (let i in mergeIds)` != `for (const id of mergeIds)`. `in` gives you the index (i) whereas `of` is a for each

Comment (martabal) [reply server/src/domain/person/person.service.ts:459]: What's the benefit over `for (let i = 0; i < mergeIds.length; i++) ` ? It's faster and you can easily have the object with the current index

Comment (martabal) [reply server/src/domain/person/person.service.ts:459]: I only use `for (const id of mergeIds)` when I don't need the index because it's easier to maintain even if it's slower

Comment (danieldietzler) [reply server/src/domain/person/person.service.ts:459]: Yeah you're right I wrote bullshit. Sorry :/

Comment (martabal) [reply server/src/domain/person/person.service.ts:459]: ahaha don't worry, I'm always curious if there's a new way to do things better :smile:

**Review by danieldietzler — APPROVED** [review]
LGTM! Thanks a lot :)

File: server/src/domain/person/person.service.ts (hunk 1 above)
Comment (jrasm91) [inline server/src/domain/person/person.service.ts:459]: Even if you need to change the loop here, at least add a second line `const mergeId = mergeIds[i];` instead of changing every subsequent line.

File: server/src/domain/person/person.service.ts [hunk 3]
```diff
@@ -442,46 +443,59 @@ export class PersonService {
...
       try {
-        const mergePerson = await this.repository.getById(mergeId);
+        const mergePerson = await this.repository.getById(mergeIds[i]);
         if (!mergePerson) {
-          results.push({ id: mergeId, success: false, error: BulkIdErrorReason.NOT_FOUND });
+          results.push({ id: mergeIds[i], success: false, error: BulkIdErrorReason.NOT_FOUND });
           continue;
         }
-
+        if (mergeIds.length === 1 && i === 0) {
```
Comment (jrasm91) [inline server/src/domain/person/person.service.ts:472]: why does it only work for the first person? Why can't, inside the for loop, you just have logic like "if person doesn't have a name or or birth day and the current merge target does, save it". I don't see why we need specialized logic that only applies for the first person.

File: server/src/domain/person/person.service.ts [hunk 4]
```diff
@@ -442,46 +443,59 @@ export class PersonService {
...
+        this.logger.error(`Unable to merge ${mergeIds[i]} into ${id}: ${error}`, error?.stack);
+        results.push({ id: mergeIds[i], success: false, error: BulkIdErrorReason.UNKNOWN });
       }
     }
-
-    return results;
+    return {
+      results,
+      person: primaryPerson,
+    };
```
Comment (jrasm91) [inline server/src/domain/person/person.service.ts:498]: I don't like this return structure very much. It makes me think this endpoint is doing too many things. I'd prefer to either:
1. Leave the code to update the person and then have the web assume the person changed and call get person by id after the merge completes, or 
2. Make two separate requests from the web: one to merge people and a second one to update the person

Basically, this looks bad to me and it would be better to split up the requests to be merge + refresh or merge + update

Comment (martabal) [reply server/src/domain/person/person.service.ts]: You're right. let's change it; I like option 1

File: server/src/domain/person/person.service.ts (hunk 3 above)
Comment (jrasm91) [reply server/src/domain/person/person.service.ts:472]: I think this is a highly specialized case and it is a bit weird to figure out when it will copy data and when it won't. Can we not just have something generic like "when merging person, if the birthdate or name is empty, it will use the first available one from the merge candidates"?

Comment (martabal) [reply server/src/domain/person/person.service.ts:472]: Sounds good

File: server/test/fixtures/person.stub.ts [hunk 5]
```diff
@@ -128,4 +128,18 @@ export const personStub = {
     faceAsset: null,
     isHidden: false,
   }),
+  RandomPerson: Object.freeze<PersonEntity>({
```
Comment (jrasm91) [inline server/test/fixtures/person.stub.ts:131]: Start with lower case r

**Review by jrasm91 — COMMENTED** [review]
Are you going to remove the restriction that this only applies to the first merge person?

**Comment by jrasm91** [comment]
I agree we should just do this by default and no need to make it something the user can turn on and off.

**Review by jrasm91 — APPROVED** [review]

**Comment by jrasm91** [comment]
I fixed it so that the merged person's data is selectively copied over. The way it was written before could allow data to be overridden in some scenarios, which seemed undesirable.
